File: frontend/assets/js/dashboard-service.js
/**
 * Dashboard Service - Singleton service for dashboard operations
 *
 * Provides comprehensive API for:
 * - Dashboard CRUD operations
 * - Dashboard page management
 * - Widget management
 * - Widget data fetching (reuses report infrastructure)
 * - Dashboard sharing and snapshots
 */

import { authService } from './auth-service.js';

class DashboardService {
    constructor() {
        if (DashboardService.instance) {
            return DashboardService.instance;
        }
        this.baseUrl = '/api/v1/dashboards';
        DashboardService.instance = this;
    }

    /**
     * Helper method to make authenticated API calls
     */
    async _fetchWithAuth(endpoint, options = {}) {
        const token = authService.getToken();
        if (!token) {
            throw new Error('Not authenticated');
        }

        const headers = {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
            ...options.headers
        };

        const response = await fetch(endpoint, {
            ...options,
            headers
        });

        if (!response.ok) {
            const error = await response.json().catch(() => ({ detail: 'Request failed' }));
            throw new Error(error.detail || `HTTP ${response.status}`);
        }

        // Handle 204 No Content
        if (response.status === 204) {
            return null;
        }

        return response.json();
    }

    // ==================== Dashboard Methods ====================

    /**
     * Create a new dashboard
     */
    async createDashboard(dashboardData) {
        return this._fetchWithAuth(`${this.baseUrl}`, {
            method: 'POST',
            body: JSON.stringify(dashboardData)
        });
    }

    /**
     * List all dashboards
     */
    async listDashboards(filters = {}) {
        const params = new URLSearchParams();
        if (filters.category) params.append('category', filters.category);
        if (filters.favorites_only) params.append('favorites_only', 'true');
        if (filters.skip) params.append('skip', filters.skip);
        if (filters.limit) params.append('limit', filters.limit);

        const url = `${this.baseUrl}${params.toString() ? '?' + params.toString() : ''}`;
        return this._fetchWithAuth(url);
    }

    /**
     * Get a specific dashboard
     */
    async getDashboard(dashboardId) {
        return this._fetchWithAuth(`${this.baseUrl}/${dashboardId}`);
    }

    /**
     * Update a dashboard
     */
    async updateDashboard(dashboardId, dashboardData) {
        return this._fetchWithAuth(`${this.baseUrl}/${dashboardId}`, {
            method: 'PUT',
            body: JSON.stringify(dashboardData)
        });
    }

    /**
     * Delete a dashboard
     */
    async deleteDashboard(dashboardId) {
        return this._fetchWithAuth(`${this.baseUrl}/${dashboardId}`, {
            method: 'DELETE'
        });
    }

    /**
     * Clone a dashboard
     */
    async cloneDashboard(dashboardId, newName) {
        return this._fetchWithAuth(`${this.baseUrl}/${dashboardId}/clone`, {
            method: 'POST',
            body: JSON.stringify({ name: newName })
        });
    }

    // ==================== Dashboard Page Methods ====================

    /**
     * Create a new page
     */
    async createPage(pageData) {
        return this._fetchWithAuth(`${this.baseUrl}/pages`, {
            method: 'POST',
            body: JSON.stringify(pageData)
        });
    }

    /**
     * Update a page
     */
    async updatePage(pageId, pageData) {
        return this._fetchWithAuth(`${this.baseUrl}/pages/${pageId}`, {
            method: 'PUT',
            body: JSON.stringify(pageData)
        });
    }

    /**
     * Delete a page
     */
    async deletePage(pageId) {
        return this._fetchWithAuth(`${this.baseUrl}/pages/${pageId}`, {
            method: 'DELETE'
        });
    }

    // ==================== Dashboard Widget Methods ====================

    /**
     * Create a new widget
     */
    async createWidget(widgetData) {
        return this._fetchWithAuth(`${this.baseUrl}/widgets`, {
            method: 'POST',
            body: JSON.stringify(widgetData)
        });
    }

    /**
     * Update a widget
     */
    async updateWidget(widgetId, widgetData) {
        return this._fetchWithAuth(`${this.baseUrl}/widgets/${widgetId}`, {
            method: 'PUT',
            body: JSON.stringify(widgetData)
        });
    }

    /**
     * Delete a widget
     */
    async deleteWidget(widgetId) {
        return this._fetchWithAuth(`${this.baseUrl}/widgets/${widgetId}`, {
            method: 'DELETE'
        });
    }

    /**
     * Bulk update widgets (for repositioning)
     */
    async bulkUpdateWidgets(updates) {
        return this._fetchWithAuth(`${this.baseUrl}/widgets/bulk-update`, {
            method: 'POST',
            body: JSON.stringify({ updates })
        });
    }

    // ==================== Widget Data Methods (Reuses Report Infrastructure) ====================

    /**
     * Get data for a widget
     * This reuses the report execution infrastructure on the backend
     */
    async getWidgetData(widgetId, parameters = {}, useCache = true) {
        return this._fetchWithAuth(`${this.baseUrl}/widgets/data`, {
            method: 'POST',
            body: JSON.stringify({
                widget_id: widgetId,
                parameters,
                use_cache: useCache
            })
        });
    }

    // ==================== Dashboard Sharing Methods ====================

    /**
     * Share a dashboard
     */
    async shareDashboard(shareData) {
        return this._fetchWithAuth(`${this.baseUrl}/shares`, {
            method: 'POST',
            body: JSON.stringify(shareData)
        });
    }

    // ==================== Dashboard Snapshot Methods ====================

    /**
     * Create a dashboard snapshot
     */
    async createSnapshot(snapshotData) {
        return this._fetchWithAuth(`${this.baseUrl}/snapshots`, {
            method: 'POST',
            body: JSON.stringify(snapshotData)
        });
    }

    // ==================== Helper Methods ====================

    /**
     * Get widget type icon
     */
    getWidgetTypeIcon(widgetType) {
        const icons = {
            'report_table': '📊',
            'chart': '📈',
            'kpi_card': '📌',
            'metric': '🔢',
            'text': '📝',
            'iframe': '🌐',
            'image': '🖼️',
            'filter_panel': '🔍'
        };
        return icons[widgetType] || '📦';
    }

    /**
     * Get widget type name
     */
    getWidgetTypeName(widgetType) {
        const names = {
            'report_table': 'Report Table',
            'chart': 'Chart',
            'kpi_card': 'KPI Card',
            'metric': 'Metric',
            'text': 'Text',
            'iframe': 'Embedded Page',
            'image': 'Image',
            'filter_panel': 'Filter Panel'
        };
        return names[widgetType] || 'Unknown';
    }

    /**
     * Validate widget position
     */
    validatePosition(position, layoutConfig = {}) {
        const columns = layoutConfig.columns || 12;
        const errors = [];

        if (position.x < 0 || position.x >= columns) {
            errors.push(`X position must be between 0 and ${columns - 1}`);
        }

        if (position.y < 0) {
            errors.push('Y position must be non-negative');
        }

        if (position.w <= 0 || position.w > columns) {
            errors.push(`Width must be between 1 and ${columns}`);
        }

        if (position.h <= 0) {
            errors.push('Height must be positive');
        }

        if (position.x + position.w > columns) {
            errors.push(`Widget extends beyond grid boundary (${columns} columns)`);
        }

        return errors;
    }

    /**
     * Calculate optimal widget position for new widget
     */
    calculateOptimalPosition(existingWidgets, defaultSize = { w: 4, h: 4 }, columns = 12) {
        // Create a grid to track occupied spaces
        const maxY = Math.max(0, ...existingWidgets.map(w => w.position.y + w.position.h));
        const grid = Array(maxY + 10).fill(null).map(() => Array(columns).fill(false));

        // Mark occupied spaces
        existingWidgets.forEach(widget => {
            const { x, y, w, h } = widget.position;
            for (let row = y; row < y + h; row++) {
                for (let col = x; col < x + w; col++) {
                    if (grid[row] && grid[row][col] !== undefined) {
                        grid[row][col] = true;
                    }
                }
            }
        });

        // Find first available spot
        for (let y = 0; y < grid.length; y++) {
            for (let x = 0; x <= columns - defaultSize.w; x++) {
                // Check if this spot can fit the widget
                let canFit = true;
                for (let row = y; row < y + defaultSize.h && row < grid.length; row++) {
                    for (let col = x; col < x + defaultSize.w; col++) {
                        if (grid[row][col]) {
                            canFit = false;
                            break;
                        }
                    }
                    if (!canFit) break;
                }

                if (canFit) {
                    return { x, y, ...defaultSize };
                }
            }
        }

        // If no spot found, add to bottom
        return { x: 0, y: maxY + 1, ...defaultSize };
    }

    /**
     * Get chart type options
     */
    getChartTypes() {
        return [
            { value: 'bar', label: 'Bar Chart', icon: '📊' },
            { value: 'line', label: 'Line Chart', icon: '📈' },
            { value: 'pie', label: 'Pie Chart', icon: '🥧' },
            { value: 'donut', label: 'Donut Chart', icon: '🍩' },
            { value: 'area', label: 'Area Chart', icon: '📉' },
            { value: 'scatter', label: 'Scatter Plot', icon: '⚫' },
            { value: 'gauge', label: 'Gauge', icon: '🎯' },
            { value: 'funnel', label: 'Funnel Chart', icon: '🔻' },
            { value: 'heatmap', label: 'Heatmap', icon: '🔥' }
        ];
    }

    /**
     * Get refresh interval options
     */
    getRefreshIntervals() {
        return [
            { value: 'none', label: 'No Auto-Refresh' },
            { value: '30s', label: 'Every 30 seconds' },
            { value: '1m', label: 'Every minute' },
            { value: '5m', label: 'Every 5 minutes' },
            { value: '15m', label: 'Every 15 minutes' },
            { value: '30m', label: 'Every 30 minutes' },
            { value: '1h', label: 'Every hour' }
        ];
    }

    /**
     * Parse refresh interval to milliseconds
     */
    parseRefreshInterval(interval) {
        const map = {
            'none': 0,
            '30s': 30 * 1000,
            '1m': 60 * 1000,
            '5m': 5 * 60 * 1000,
            '15m': 15 * 60 * 1000,
            '30m': 30 * 60 * 1000,
            '1h': 60 * 60 * 1000
        };
        return map[interval] || 0;
    }
}

// Create singleton instance
export const dashboardService = new DashboardService();

File: backend/app/services/dashboard_service.py
"""
Dashboard service - business logic for dashboards.

Reuses ReportService for data fetching and execution.
"""
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
import hashlib
import json

from app.models.dashboard import (
    Dashboard,
    DashboardPage,
    DashboardWidget,
    DashboardShare,
    DashboardSnapshot,
    WidgetDataCache,
    WidgetType
)
from app.schemas.dashboard import (
    DashboardCreate,
    DashboardUpdate,
    DashboardPageCreate,
    DashboardPageUpdate,
    DashboardWidgetCreate,
    DashboardWidgetUpdate,
    DashboardShareCreate,
    DashboardSnapshotCreate
)
from app.services.report_service import ReportService
from app.schemas.report import ReportExecutionRequest


class DashboardService:
    """Service for dashboard operations."""

    # ==================== Dashboard CRUD ====================

    @staticmethod
    def create_dashboard(
        db: Session,
        tenant_id: int,
        user_id: int,
        dashboard_data: DashboardCreate
    ) -> Dashboard:
        """Create a new dashboard."""
        dashboard_dict = dashboard_data.model_dump()

        # Convert Pydantic models to dict for JSON columns
        json_fields = ['global_parameters', 'global_filters', 'tags', 'allowed_roles', 'allowed_users']
        for field in json_fields:
            if field in dashboard_dict and dashboard_dict[field] is not None:
                if isinstance(dashboard_dict[field], list):
                    dashboard_dict[field] = [
                        item.model_dump() if hasattr(item, 'model_dump') else item
                        for item in dashboard_dict[field]
                    ]

        db_dashboard = Dashboard(
            tenant_id=tenant_id,
            created_by=user_id,
            **dashboard_dict
        )
        db.add(db_dashboard)
        db.commit()
        db.refresh(db_dashboard)
        return db_dashboard

    @staticmethod
    def get_dashboard(
        db: Session,
        tenant_id: int,
        dashboard_id: int,
        user_id: Optional[int] = None
    ) -> Optional[Dashboard]:
        """Get dashboard by ID with permission check."""
        dashboard = db.query(Dashboard).filter(
            Dashboard.id == dashboard_id,
            Dashboard.tenant_id == tenant_id,
            Dashboard.is_active == True
        ).first()

        if not dashboard:
            return None

        # Check permissions
        if not dashboard.is_public and user_id:
            if dashboard.allowed_users and user_id not in dashboard.allowed_users:
                # Check if user has access via share
                share = db.query(DashboardShare).filter(
                    DashboardShare.dashboard_id == dashboard_id,
                    DashboardShare.shared_with_user_id == user_id,
                    DashboardShare.can_view == True
                ).first()
                if not share:
                    return None

        return dashboard

    @staticmethod
    def list_dashboards(
        db: Session,
        tenant_id: int,
        user_id: Optional[int] = None,
        category: Optional[str] = None,
        favorites_only: bool = False,
        skip: int = 0,
        limit: int = 100
    ) -> List[Dashboard]:
        """List dashboards with filtering."""
        query = db.query(Dashboard).filter(
            Dashboard.tenant_id == tenant_id,
            Dashboard.is_active == True
        )

        if category:
            query = query.filter(Dashboard.category == category)

        if favorites_only:
            query = query.filter(Dashboard.is_favorite == True)

        # Filter by permissions
        if user_id:
            query = query.filter(
                or_(
                    Dashboard.is_public == True,
                    Dashboard.created_by == user_id,
                    Dashboard.allowed_users.contains([user_id])
                )
            )

        return query.offset(skip).limit(limit).all()

    @staticmethod
    def update_dashboard(
        db: Session,
        tenant_id: int,
        dashboard_id: int,
        dashboard_data: DashboardUpdate
    ) -> Optional[Dashboard]:
        """Update dashboard."""
        db_dashboard = db.query(Dashboard).filter(
            Dashboard.id == dashboard_id,
            Dashboard.tenant_id == tenant_id
        ).first()

        if not db_dashboard:
            return None

        update_data = dashboard_data.model_dump(exclude_unset=True)

        # Convert Pydantic models to dict for JSON columns
        json_fields = ['global_parameters', 'global_filters', 'tags', 'allowed_roles', 'allowed_users']
        for field in json_fields:
            if field in update_data and update_data[field] is not None:
                if isinstance(update_data[field], list):
                    update_data[field] = [
                        item.model_dump() if hasattr(item, 'model_dump') else item
                        for item in update_data[field]
                    ]

        for field, value in update_data.items():
            setattr(db_dashboard, field, value)

        db_dashboard.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(db_dashboard)
        return db_dashboard

    @staticmethod
    def delete_dashboard(
        db: Session,
        tenant_id: int,
        dashboard_id: int
    ) -> bool:
        """Soft delete dashboard."""
        db_dashboard = db.query(Dashboard).filter(
            Dashboard.id == dashboard_id,
            Dashboard.tenant_id == tenant_id
        ).first()

        if not db_dashboard:
            return False

        db_dashboard.is_active = False
        db.commit()
        return True

    @staticmethod
    def clone_dashboard(
        db: Session,
        tenant_id: int,
        user_id: int,
        dashboard_id: int,
        new_name: str
    ) -> Optional[Dashboard]:
        """Clone an existing dashboard."""
        source_dashboard = DashboardService.get_dashboard(db, tenant_id, dashboard_id, user_id)
        if not source_dashboard:
            return None

        # Create new dashboard
        new_dashboard = Dashboard(
            tenant_id=tenant_id,
            created_by=user_id,
            name=new_name,
            description=source_dashboard.description,
            category=source_dashboard.category,
            tags=source_dashboard.tags,
            layout_type=source_dashboard.layout_type,
            theme=source_dashboard.theme,
            global_parameters=source_dashboard.global_parameters,
            global_filters=source_dashboard.global_filters,
            refresh_interval=source_dashboard.refresh_interval,
            show_header=source_dashboard.show_header,
            show_filters=source_dashboard.show_filters
        )
        db.add(new_dashboard)
        db.flush()

        # Clone pages and widgets
        for page in source_dashboard.pages:
            new_page = DashboardPage(
                dashboard_id=new_dashboard.id,
                tenant_id=tenant_id,
                name=page.name,
                slug=page.slug,
                description=page.description,
                icon=page.icon,
                layout_config=page.layout_config,
                order=page.order,
                is_default=page.is_default
            )
            db.add(new_page)
            db.flush()

            # Clone widgets
            for widget in page.widgets:
                new_widget = DashboardWidget(
                    page_id=new_page.id,
                    tenant_id=tenant_id,
                    title=widget.title,
                    description=widget.description,
                    widget_type=widget.widget_type,
                    report_definition_id=widget.report_definition_id,
                    data_source_config=widget.data_source_config,
                    widget_config=widget.widget_config,
                    chart_config=widget.chart_config,
                    filter_mapping=widget.filter_mapping,
                    position=widget.position,
                    order=widget.order,
                    show_title=widget.show_title,
                    show_border=widget.show_border,
                    background_color=widget.background_color,
                    auto_refresh=widget.auto_refresh,
                    refresh_interval=widget.refresh_interval
                )
                db.add(new_widget)

        db.commit()
        db.refresh(new_dashboard)
        return new_dashboard

    # ==================== Dashboard Page CRUD ====================

    @staticmethod
    def create_page(
        db: Session,
        tenant_id: int,
        page_data: DashboardPageCreate
    ) -> DashboardPage:
        """Create a new dashboard page."""
        page_dict = page_data.model_dump()

        # Convert layout_config if it's a Pydantic model
        if 'layout_config' in page_dict and page_dict['layout_config'] is not None:
            if hasattr(page_dict['layout_config'], 'model_dump'):
                page_dict['layout_config'] = page_dict['layout_config'].model_dump()

        db_page = DashboardPage(
            tenant_id=tenant_id,
            **page_dict
        )
        db.add(db_page)
        db.commit()
        db.refresh(db_page)
        return db_page

    @staticmethod
    def update_page(
        db: Session,
        tenant_id: int,
        page_id: int,
        page_data: DashboardPageUpdate
    ) -> Optional[DashboardPage]:
        """Update dashboard page."""
        db_page = db.query(DashboardPage).filter(
            DashboardPage.id == page_id,
            DashboardPage.tenant_id == tenant_id
        ).first()

        if not db_page:
            return None

        update_data = page_data.model_dump(exclude_unset=True)

        # Convert layout_config if it's a Pydantic model
        if 'layout_config' in update_data and update_data['layout_config'] is not None:
            if hasattr(update_data['layout_config'], 'model_dump'):
                update_data['layout_config'] = update_data['layout_config'].model_dump()

        for field, value in update_data.items():
            setattr(db_page, field, value)

        db_page.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(db_page)
        return db_page

    @staticmethod
    def delete_page(
        db: Session,
        tenant_id: int,
        page_id: int
    ) -> bool:
        """Delete dashboard page."""
        db_page = db.query(DashboardPage).filter(
            DashboardPage.id == page_id,
            DashboardPage.tenant_id == tenant_id
        ).first()

        if not db_page:
            return False

        db.delete(db_page)
        db.commit()
        return True

    # ==================== Dashboard Widget CRUD ====================

    @staticmethod
    def create_widget(
        db: Session,
        tenant_id: int,
        widget_data: DashboardWidgetCreate
    ) -> DashboardWidget:
        """Create a new dashboard widget."""
        widget_dict = widget_data.model_dump()

        # Convert Pydantic models to dict for JSON columns
        json_fields = ['data_source_config', 'widget_config', 'chart_config',
                      'filter_mapping', 'position']
        for field in json_fields:
            if field in widget_dict and widget_dict[field] is not None:
                if hasattr(widget_dict[field], 'model_dump'):
                    widget_dict[field] = widget_dict[field].model_dump()

        db_widget = DashboardWidget(
            tenant_id=tenant_id,
            **widget_dict
        )
        db.add(db_widget)
        db.commit()
        db.refresh(db_widget)
        return db_widget

    @staticmethod
    def update_widget(
        db: Session,
        tenant_id: int,
        widget_id: int,
        widget_data: DashboardWidgetUpdate
    ) -> Optional[DashboardWidget]:
        """Update dashboard widget."""
        db_widget = db.query(DashboardWidget).filter(
            DashboardWidget.id == widget_id,
            DashboardWidget.tenant_id == tenant_id
        ).first()

        if not db_widget:
            return None

        update_data = widget_data.model_dump(exclude_unset=True)

        # Convert Pydantic models to dict for JSON columns
        json_fields = ['data_source_config', 'widget_config', 'chart_config',
                      'filter_mapping', 'position']
        for field in json_fields:
            if field in update_data and update_data[field] is not None:
                if hasattr(update_data[field], 'model_dump'):
                    update_data[field] = update_data[field].model_dump()

        for field, value in update_data.items():
            setattr(db_widget, field, value)

        db_widget.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(db_widget)
        return db_widget

    @staticmethod
    def delete_widget(
        db: Session,
        tenant_id: int,
        widget_id: int
    ) -> bool:
        """Delete dashboard widget."""
        db_widget = db.query(DashboardWidget).filter(
            DashboardWidget.id == widget_id,
            DashboardWidget.tenant_id == tenant_id
        ).first()

        if not db_widget:
            return False

        db.delete(db_widget)
        db.commit()
        return True

    @staticmethod
    def bulk_update_widgets(
        db: Session,
        tenant_id: int,
        updates: List[Dict[str, Any]]
    ) -> bool:
        """Bulk update widget positions and order."""
        for update in updates:
            widget_id = update.get('widget_id')
            position = update.get('position')
            order = update.get('order')

            db_widget = db.query(DashboardWidget).filter(
                DashboardWidget.id == widget_id,
                DashboardWidget.tenant_id == tenant_id
            ).first()

            if db_widget:
                if position:
                    db_widget.position = position
                if order is not None:
                    db_widget.order = order
                db_widget.updated_at = datetime.utcnow()

        db.commit()
        return True

    # ==================== Widget Data Retrieval (Reuses ReportService) ====================

    @staticmethod
    def get_widget_data(
        db: Session,
        tenant_id: int,
        user_id: int,
        widget_id: int,
        parameters: Optional[Dict[str, Any]] = None,
        use_cache: bool = True
    ) -> Dict[str, Any]:
        """
        Get data for a widget - REUSES ReportService for report-based widgets.
        """
        widget = db.query(DashboardWidget).filter(
            DashboardWidget.id == widget_id,
            DashboardWidget.tenant_id == tenant_id
        ).first()

        if not widget:
            raise ValueError("Widget not found")

        # Check cache first
        if use_cache:
            cached_data = DashboardService._get_cached_widget_data(
                db, tenant_id, widget_id, parameters
            )
            if cached_data:
                return {
                    'data': cached_data,
                    'cached': True,
                    'widget_type': widget.widget_type
                }

        # Get data based on widget type
        data = None
        execution_time_ms = 0
        start_time = datetime.utcnow()

        if widget.widget_type == WidgetType.REPORT_TABLE and widget.report_definition_id:
            # REUSE ReportService for report-based widgets
            execution_request = ReportExecutionRequest(
                report_definition_id=widget.report_definition_id,
                parameters=parameters or {},
                use_cache=use_cache
            )
            execution = ReportService.execute_report(
                db, tenant_id, user_id, execution_request
            )

            if execution.status == 'completed':
                # Fetch the actual data from the execution
                query_result = ReportService._build_and_execute_query(
                    db,
                    tenant_id,
                    ReportService.get_report_definition(
                        db, tenant_id, widget.report_definition_id, user_id
                    ),
                    parameters or {}
                )
                data = query_result['data']
                execution_time_ms = execution.execution_time_ms or 0

        elif widget.widget_type == WidgetType.CHART and widget.report_definition_id:
            # REUSE ReportService for chart data
            report_def = ReportService.get_report_definition(
                db, tenant_id, widget.report_definition_id, user_id
            )
            if report_def:
                query_result = ReportService._build_and_execute_query(
                    db, tenant_id, report_def, parameters or {}
                )
                data = DashboardService._transform_data_for_chart(
                    query_result['data'],
                    widget.chart_config
                )

        elif widget.widget_type == WidgetType.KPI_CARD:
            # For KPI cards, fetch single metric from report
            if widget.report_definition_id:
                report_def = ReportService.get_report_definition(
                    db, tenant_id, widget.report_definition_id, user_id
                )
                if report_def:
                    query_result = ReportService._build_and_execute_query(
                        db, tenant_id, report_def, parameters or {}
                    )
                    data = DashboardService._transform_data_for_kpi(
                        query_result['data'],
                        widget.widget_config
                    )

        elif widget.widget_type == WidgetType.TEXT:
            # Static text widget
            data = widget.widget_config

        elif widget.widget_type == WidgetType.IFRAME:
            # Iframe widget
            data = widget.widget_config

        end_time = datetime.utcnow()
        if execution_time_ms == 0:
            execution_time_ms = int((end_time - start_time).total_seconds() * 1000)

        # Cache the result
        if use_cache and data:
            DashboardService._cache_widget_data(
                db, tenant_id, widget_id, parameters, data
            )

        return {
            'data': data,
            'cached': False,
            'execution_time_ms': execution_time_ms,
            'widget_type': widget.widget_type
        }

    @staticmethod
    def _transform_data_for_chart(data: List[Dict], chart_config: Optional[Dict]) -> Dict:
        """Transform report data for chart visualization."""
        if not chart_config:
            return {'data': data}

        # Extract chart configuration
        x_axis = chart_config.get('x_axis')
        y_axis = chart_config.get('y_axis', [])

        # Build chart-friendly format
        chart_data = {
            'labels': [row.get(x_axis) for row in data],
            'datasets': []
        }

        for y_field in y_axis:
            chart_data['datasets'].append({
                'label': y_field,
                'data': [row.get(y_field) for row in data]
            })

        return chart_data

    @staticmethod
    def _transform_data_for_kpi(data: List[Dict], widget_config: Optional[Dict]) -> Dict:
        """Transform report data for KPI card."""
        if not widget_config or not data:
            return {'value': 0}

        value_field = widget_config.get('value_field')

        # Aggregate if multiple rows
        if len(data) > 0:
            if len(data) == 1:
                value = data[0].get(value_field, 0)
            else:
                # Sum all values
                value = sum(row.get(value_field, 0) for row in data)
        else:
            value = 0

        return {
            'value': value,
            'label': widget_config.get('label', ''),
            'format': widget_config.get('format', 'number')
        }

    @staticmethod
    def _get_cached_widget_data(
        db: Session,
        tenant_id: int,
        widget_id: int,
        parameters: Optional[Dict[str, Any]]
    ) -> Optional[Any]:
        """Get cached widget data."""
        cache_key = DashboardService._generate_widget_cache_key(widget_id, parameters)

        cache_entry = db.query(WidgetDataCache).filter(
            WidgetDataCache.tenant_id == tenant_id,
            WidgetDataCache.widget_id == widget_id,
            WidgetDataCache.cache_key == cache_key,
            WidgetDataCache.expires_at > datetime.utcnow()
        ).first()

        if cache_entry:
            cache_entry.hit_count += 1
            db.commit()
            return cache_entry.cached_data

        return None

    @staticmethod
    def _cache_widget_data(
        db: Session,
        tenant_id: int,
        widget_id: int,
        parameters: Optional[Dict[str, Any]],
        data: Any,
        ttl_minutes: int = 15
    ):
        """Cache widget data."""
        cache_key = DashboardService._generate_widget_cache_key(widget_id, parameters)
        params_hash = hashlib.sha256(
            json.dumps(parameters or {}, sort_keys=True).encode()
        ).hexdigest()

        # Check if cache entry exists
        cache_entry = db.query(WidgetDataCache).filter(
            WidgetDataCache.cache_key == cache_key
        ).first()

        expires_at = datetime.utcnow() + timedelta(minutes=ttl_minutes)

        if cache_entry:
            cache_entry.cached_data = data
            cache_entry.expires_at = expires_at
        else:
            cache_entry = WidgetDataCache(
                widget_id=widget_id,
                tenant_id=tenant_id,
                cache_key=cache_key,
                parameters_hash=params_hash,
                cached_data=data,
                expires_at=expires_at
            )
            db.add(cache_entry)

        db.commit()

    @staticmethod
    def _generate_widget_cache_key(widget_id: int, parameters: Optional[Dict[str, Any]]) -> str:
        """Generate cache key for widget data."""
        params_str = json.dumps(parameters or {}, sort_keys=True)
        return f"widget_{widget_id}_{hashlib.sha256(params_str.encode()).hexdigest()}"

    # ==================== Dashboard Sharing ====================

    @staticmethod
    def create_share(
        db: Session,
        tenant_id: int,
        user_id: int,
        share_data: DashboardShareCreate
    ) -> DashboardShare:
        """Create dashboard share."""
        import secrets

        db_share = DashboardShare(
            tenant_id=tenant_id,
            created_by=user_id,
            share_token=secrets.token_urlsafe(32) if not share_data.shared_with_user_id else None,
            **share_data.model_dump()
        )
        db.add(db_share)
        db.commit()
        db.refresh(db_share)
        return db_share

    # ==================== Dashboard Snapshots ====================

    @staticmethod
    def create_snapshot(
        db: Session,
        tenant_id: int,
        user_id: int,
        snapshot_data: DashboardSnapshotCreate
    ) -> DashboardSnapshot:
        """Create dashboard snapshot."""
        # Get full dashboard data
        dashboard = DashboardService.get_dashboard(
            db, tenant_id, snapshot_data.dashboard_id, user_id
        )

        if not dashboard:
            raise ValueError("Dashboard not found")

        # Serialize dashboard state
        snapshot_dict = {
            'dashboard': {
                'id': dashboard.id,
                'name': dashboard.name,
                'description': dashboard.description,
                'layout_type': dashboard.layout_type,
                'theme': dashboard.theme
            },
            'pages': [],
            'created_at': datetime.utcnow().isoformat()
        }

        for page in dashboard.pages:
            page_dict = {
                'id': page.id,
                'name': page.name,
                'widgets': []
            }
            for widget in page.widgets:
                widget_dict = {
                    'id': widget.id,
                    'title': widget.title,
                    'widget_type': widget.widget_type,
                    'position': widget.position
                }
                page_dict['widgets'].append(widget_dict)
            snapshot_dict['pages'].append(page_dict)

        db_snapshot = DashboardSnapshot(
            tenant_id=tenant_id,
            created_by=user_id,
            dashboard_id=snapshot_data.dashboard_id,
            name=snapshot_data.name,
            description=snapshot_data.description,
            snapshot_data=snapshot_dict,
            parameters_used=snapshot_data.parameters_used
        )
        db.add(db_snapshot)
        db.commit()
        db.refresh(db_snapshot)
        return db_snapshot
